{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a6308833",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import sys\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "72a18fc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Clade_0</th>\n",
       "      <th>Clade_1</th>\n",
       "      <th>Clade_2</th>\n",
       "      <th>Clade_3</th>\n",
       "      <th>Clade</th>\n",
       "      <th>Ecotype</th>\n",
       "      <th>Allele</th>\n",
       "      <th>File</th>\n",
       "      <th>HV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ZM00029AB319340_P001</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>ki3</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ZM00023AB313710_P001</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>cml247</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ZM00042AB469660_P001</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>tzi8</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ZM00001EB304860_P001</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>b73</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ZM00022AB310170_P001</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int3448_22</td>\n",
       "      <td>cml228</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3491</td>\n",
       "      <td>ZM00030AB305770_P001</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>ki11</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3492</td>\n",
       "      <td>ZM00032AB317850_P001</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>m37w</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3493</td>\n",
       "      <td>ZM00042AB318810_P001</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>tzi8</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3494</td>\n",
       "      <td>ZM00039AB310760_P001</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>oh43</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3495</td>\n",
       "      <td>ZM00028AB312350_P001</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Int6844_19</td>\n",
       "      <td>il14h</td>\n",
       "      <td>1</td>\n",
       "      <td>RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3496 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene     Clade_0     Clade_1 Clade_2 Clade_3  \\\n",
       "0     ZM00029AB319340_P001  Int3448_22  Int3448_22     NaN     NaN   \n",
       "1     ZM00023AB313710_P001  Int3448_22  Int3448_22     NaN     NaN   \n",
       "2     ZM00042AB469660_P001  Int3448_22  Int3448_22     NaN     NaN   \n",
       "3     ZM00001EB304860_P001  Int3448_22  Int3448_22     NaN     NaN   \n",
       "4     ZM00022AB310170_P001  Int3448_22  Int3448_22     NaN     NaN   \n",
       "...                    ...         ...         ...     ...     ...   \n",
       "3491  ZM00030AB305770_P001  Int6844_19  Int6844_19     NaN     NaN   \n",
       "3492  ZM00032AB317850_P001  Int6844_19  Int6844_19     NaN     NaN   \n",
       "3493  ZM00042AB318810_P001  Int6844_19  Int6844_19     NaN     NaN   \n",
       "3494  ZM00039AB310760_P001  Int6844_19  Int6844_19     NaN     NaN   \n",
       "3495  ZM00028AB312350_P001  Int6844_19  Int6844_19     NaN     NaN   \n",
       "\n",
       "           Clade Ecotype  Allele  \\\n",
       "0     Int3448_22     ki3       1   \n",
       "1     Int3448_22  cml247       1   \n",
       "2     Int3448_22    tzi8       1   \n",
       "3     Int3448_22     b73       1   \n",
       "4     Int3448_22  cml228       1   \n",
       "...          ...     ...     ...   \n",
       "3491  Int6844_19    ki11       1   \n",
       "3492  Int6844_19    m37w       1   \n",
       "3493  Int6844_19    tzi8       1   \n",
       "3494  Int6844_19    oh43       1   \n",
       "3495  Int6844_19   il14h       1   \n",
       "\n",
       "                                                   File  HV  \n",
       "0     RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa   0  \n",
       "1     RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa   0  \n",
       "2     RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa   0  \n",
       "3     RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa   0  \n",
       "4     RAxML_tree_pbNB-ARC/Initial_Clades/Int3448_22.afa   0  \n",
       "...                                                 ...  ..  \n",
       "3491  RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa   0  \n",
       "3492  RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa   0  \n",
       "3493  RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa   0  \n",
       "3494  RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa   0  \n",
       "3495  RAxML_tree_pbNB-ARC/Initial_Clades/Int6844_19.afa   0  \n",
       "\n",
       "[3496 rows x 10 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import the gene table, which has the NLRs for all accessions \n",
    "gene_table='/global/home/users/chandlersutherland/e16/Maize_NLRome_GeneTable.txt'\n",
    "pd.read_csv(gene_table, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e6c4ac2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#glob all the gff3 files across all genomes \n",
    "gff3_files=glob.glob('/global/scratch/users/chandlersutherland/e16/*/genome/*1.gff3')\n",
    "\n",
    "#glob the fulldata.txt files \n",
    "fulldata=glob.glob('/global/scratch/users/chandlersutherland/e16/*/genome/*fulldata.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "17d40853",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gff3_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "db2fc26c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>sequence_ontology</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>308452471</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=1;Name=chromosome:Zm-B73-REFERENCE-NAM-5.0:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NAM</td>\n",
       "      <td>gene</td>\n",
       "      <td>34617</td>\n",
       "      <td>40204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Zm00001eb000010;biotype=protein_coding;logi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NAM</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>34617</td>\n",
       "      <td>40204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Zm00001eb000010_T001;Parent=Zm00001eb000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NAM</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>34617</td>\n",
       "      <td>34721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=Zm00001eb000010_T001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NAM</td>\n",
       "      <td>exon</td>\n",
       "      <td>34617</td>\n",
       "      <td>35318</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=Zm00001eb000010_T001;Name=Zm00001eb0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1143798</td>\n",
       "      <td>scaf_695</td>\n",
       "      <td>NAM</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>2336</td>\n",
       "      <td>4226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=Zm00001eb443030_T001;Parent=Zm00001eb443030...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1143799</td>\n",
       "      <td>scaf_695</td>\n",
       "      <td>NAM</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>4218</td>\n",
       "      <td>4226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=Zm00001eb443030_T001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1143800</td>\n",
       "      <td>scaf_695</td>\n",
       "      <td>NAM</td>\n",
       "      <td>exon</td>\n",
       "      <td>2336</td>\n",
       "      <td>4226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=Zm00001eb443030_T001;Name=Zm00001eb4430...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1143801</td>\n",
       "      <td>scaf_695</td>\n",
       "      <td>NAM</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2898</td>\n",
       "      <td>4217</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=Zm00001eb443030_P001;Parent=Zm00001eb443030...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1143802</td>\n",
       "      <td>scaf_695</td>\n",
       "      <td>NAM</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>2336</td>\n",
       "      <td>2897</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=Zm00001eb443030_T001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1143803 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              chr    source sequence_ontology  start        end score strand  \\\n",
       "0            chr1  assembly        chromosome      1  308452471     .      .   \n",
       "1            chr1       NAM              gene  34617      40204     .      +   \n",
       "2            chr1       NAM              mRNA  34617      40204     .      +   \n",
       "3            chr1       NAM    five_prime_UTR  34617      34721     .      +   \n",
       "4            chr1       NAM              exon  34617      35318     .      +   \n",
       "...           ...       ...               ...    ...        ...   ...    ...   \n",
       "1143798  scaf_695       NAM              mRNA   2336       4226     .      -   \n",
       "1143799  scaf_695       NAM    five_prime_UTR   4218       4226     .      -   \n",
       "1143800  scaf_695       NAM              exon   2336       4226     .      -   \n",
       "1143801  scaf_695       NAM               CDS   2898       4217     .      -   \n",
       "1143802  scaf_695       NAM   three_prime_UTR   2336       2897     .      -   \n",
       "\n",
       "        phase                                         attributes  \n",
       "0           .  ID=1;Name=chromosome:Zm-B73-REFERENCE-NAM-5.0:...  \n",
       "1           .  ID=Zm00001eb000010;biotype=protein_coding;logi...  \n",
       "2           .  ID=Zm00001eb000010_T001;Parent=Zm00001eb000010...  \n",
       "3           .                        Parent=Zm00001eb000010_T001  \n",
       "4           .  Parent=Zm00001eb000010_T001;Name=Zm00001eb0000...  \n",
       "...       ...                                                ...  \n",
       "1143798     .  ID=Zm00001eb443030_T001;Parent=Zm00001eb443030...  \n",
       "1143799     .                        Parent=Zm00001eb443030_T001  \n",
       "1143800     .  Parent=Zm00001eb443030_T001;Name=Zm00001eb4430...  \n",
       "1143801     0  ID=Zm00001eb443030_P001;Parent=Zm00001eb443030...  \n",
       "1143802     .                        Parent=Zm00001eb443030_T001  \n",
       "\n",
       "[1143803 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b73='/global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3'\n",
    "gff3=pd.read_csv(b73, sep='\\t', skiprows=3, header=0, lineterminator='\\n')\n",
    "gff3=gff3.reset_index()\n",
    "gff3=gff3.rename(columns={gff3.columns[0]:'chr', gff3.columns[1]:'source', gff3.columns[2]:'sequence_ontology', gff3.columns[3]:'start', gff3.columns[4]:'end', gff3.columns[5]:'score', gff3.columns[6]:'strand', gff3.columns[7]:'phase', gff3.columns[8]:'attributes'})\n",
    "gff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d20c5c36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['logic_name=cshl_gene'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes=gff3[gff3['sequence_ontology']=='gene']\n",
    "genes['attributes'].str.split(';', expand=True)[2].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "82100994",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define a function that takes a gff3 file path as input, and outputs an all gene bed file and an NLR specific bed file. \n",
    "def bed_generator(gff_filepath, gene_table):\n",
    "    #read in and cleanup gff3 file \n",
    "    gff3=pd.read_csv(gff_filepath, sep='\\t', skiprows=3, header=0, lineterminator='\\n')\n",
    "    gff3=gff3.reset_index()\n",
    "    gff3=gff3.rename(columns={gff3.columns[0]:'chr', gff3.columns[1]:'source', gff3.columns[2]:'sequence_ontology', gff3.columns[3]:'start', gff3.columns[4]:'end', gff3.columns[5]:'score', gff3.columns[6]:'strand', gff3.columns[7]:'phase', gff3.columns[8]:'attributes'})\n",
    "\n",
    "    #get metadata\n",
    "    accession=gff_filepath.split('/')[6]\n",
    "    basename=gff_filepath.split('/')[8].split('.')[0]\n",
    "    print(accession+ ' gff3 file loaded')\n",
    "    \n",
    "    #filter gff3 to just genes, and expand attributes  \n",
    "    genes = gff3[gff3['sequence_ontology'] == 'gene']\n",
    "    attributes=genes.iloc[:,8].str.split(';',expand=True)\n",
    "    genes.loc[:,'ID']=attributes.iloc[:,0].str.split('=', expand=True)[1]\n",
    "    genes.loc[:,'biotype']=attributes.iloc[:,1].str.split('=', expand=True)[1]\n",
    "    genes.loc[:,'logic_name']=attributes.iloc[:,2].str.split('=', expand=True)[1]\n",
    "    \n",
    "    #generate a bed file for all genes \n",
    "    bed = genes.loc[:,['chr', 'start', 'end', 'ID', 'strand']]\n",
    "    bed = bed.rename(columns={'chr':'chrom', 'start':'chromStart', 'end':'chromEnd', 'ID':'name', 'strand':'strand'})\n",
    "    bed.to_csv('/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_all_gene.bed', sep='\\t')\n",
    "    print('all gene bed written to '+'/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_all_gene.bed')\n",
    "    \n",
    "    #Load NLR gene names\n",
    "    gene=pd.read_csv(gene_table,sep = '\\t')\n",
    "    nlr_df=gene[gene['Ecotype'] == accession.lower()]\n",
    "    nlrs=nlr_df.loc[:,'Gene'].str.split('_', expand=True).iloc[:,0].str.replace('ZM', 'Zm').str.replace('AB', 'ab').str.replace('EB','eb').unique()\n",
    "    \n",
    "    #quality stats \n",
    "    match=bed['name'].isin(nlrs).sum()\n",
    "    unmatch=len(nlrs)-match \n",
    "    print(str(match)+' NLRs found in gff file for '+accession+'. We are missing '+str(unmatch)+'.')\n",
    "    \n",
    "    #write out NLR bed file\n",
    "    nlr_bed=bed[bed['name'].isin(nlrs)]\n",
    "    nlr_bed.to_csv('/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_NLR.bed', sep='\\t')\n",
    "    print('NLR bed written to '+'/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_NLR.bed')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "07b7dd43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B97 gff3 file loaded\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/B97/genome/Zm-B97-REFERENCE-NAM-1_all_gene.bed\n",
      "125 NLRs found in gff file for B97. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B97/genome/Zm-B97-REFERENCE-NAM-1_NLR.bed\n",
      "Oh43 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Oh43/genome/Zm-Oh43-REFERENCE-NAM-1_all_gene.bed\n",
      "131 NLRs found in gff file for Oh43. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh43/genome/Zm-Oh43-REFERENCE-NAM-1_NLR.bed\n",
      "Tx303 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Tx303/genome/Zm-Tx303-REFERENCE-NAM-1_all_gene.bed\n",
      "121 NLRs found in gff file for Tx303. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tx303/genome/Zm-Tx303-REFERENCE-NAM-1_NLR.bed\n",
      "CML69 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML69/genome/Zm-CML69-REFERENCE-NAM-1_all_gene.bed\n",
      "133 NLRs found in gff file for CML69. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML69/genome/Zm-CML69-REFERENCE-NAM-1_NLR.bed\n",
      "CML333 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML333/genome/Zm-CML333-REFERENCE-NAM-1_all_gene.bed\n",
      "131 NLRs found in gff file for CML333. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML333/genome/Zm-CML333-REFERENCE-NAM-1_NLR.bed\n",
      "NC358 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/NC358/genome/Zm-NC358-REFERENCE-NAM-1_all_gene.bed\n",
      "132 NLRs found in gff file for NC358. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC358/genome/Zm-NC358-REFERENCE-NAM-1_NLR.bed\n",
      "NC350 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/NC350/genome/Zm-NC350-REFERENCE-NAM-1_all_gene.bed\n",
      "136 NLRs found in gff file for NC350. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC350/genome/Zm-NC350-REFERENCE-NAM-1_NLR.bed\n",
      "Mo18W gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Mo18W/genome/Zm-Mo18W-REFERENCE-NAM-1_all_gene.bed\n",
      "117 NLRs found in gff file for Mo18W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Mo18W/genome/Zm-Mo18W-REFERENCE-NAM-1_NLR.bed\n",
      "CML103 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML103/genome/Zm-CML103-REFERENCE-NAM-1_all_gene.bed\n",
      "119 NLRs found in gff file for CML103. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML103/genome/Zm-CML103-REFERENCE-NAM-1_NLR.bed\n",
      "CML247 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML247/genome/Zm-CML247-REFERENCE-NAM-1_all_gene.bed\n",
      "114 NLRs found in gff file for CML247. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML247/genome/Zm-CML247-REFERENCE-NAM-1_NLR.bed\n",
      "B73 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_all_gene.bed\n",
      "0 NLRs found in gff file for B73. We are missing 126.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_NLR.bed\n",
      "CML228 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML228/genome/Zm-CML228-REFERENCE-NAM-1_all_gene.bed\n",
      "126 NLRs found in gff file for CML228. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML228/genome/Zm-CML228-REFERENCE-NAM-1_NLR.bed\n",
      "P39 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/P39/genome/Zm-P39-REFERENCE-NAM-1_all_gene.bed\n",
      "118 NLRs found in gff file for P39. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/P39/genome/Zm-P39-REFERENCE-NAM-1_NLR.bed\n",
      "Oh7B gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Oh7B/genome/Zm-Oh7B-REFERENCE-NAM-1_all_gene.bed\n",
      "121 NLRs found in gff file for Oh7B. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh7B/genome/Zm-Oh7B-REFERENCE-NAM-1_NLR.bed\n",
      "MS71 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/MS71/genome/Zm-Ms71-REFERENCE-NAM-1_all_gene.bed\n",
      "122 NLRs found in gff file for MS71. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/MS71/genome/Zm-Ms71-REFERENCE-NAM-1_NLR.bed\n",
      "IL14H gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/IL14H/genome/Zm-Il14H-REFERENCE-NAM-1_all_gene.bed\n",
      "125 NLRs found in gff file for IL14H. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/IL14H/genome/Zm-Il14H-REFERENCE-NAM-1_NLR.bed\n",
      "Ki11 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Ki11/genome/Zm-Ki11-REFERENCE-NAM-1_all_gene.bed\n",
      "120 NLRs found in gff file for Ki11. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki11/genome/Zm-Ki11-REFERENCE-NAM-1_NLR.bed\n",
      "CML52 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML52/genome/Zm-CML52-REFERENCE-NAM-1_all_gene.bed\n",
      "130 NLRs found in gff file for CML52. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML52/genome/Zm-CML52-REFERENCE-NAM-1_NLR.bed\n",
      "CML322 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML322/genome/Zm-CML322-REFERENCE-NAM-1_all_gene.bed\n",
      "121 NLRs found in gff file for CML322. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML322/genome/Zm-CML322-REFERENCE-NAM-1_NLR.bed\n",
      "KY21 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/KY21/genome/Zm-Ky21-REFERENCE-NAM-1_all_gene.bed\n",
      "126 NLRs found in gff file for KY21. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/KY21/genome/Zm-Ky21-REFERENCE-NAM-1_NLR.bed\n",
      "M37W gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/M37W/genome/Zm-M37W-REFERENCE-NAM-1_all_gene.bed\n",
      "134 NLRs found in gff file for M37W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M37W/genome/Zm-M37W-REFERENCE-NAM-1_NLR.bed\n",
      "Tzi8 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Tzi8/genome/Zm-Tzi8-REFERENCE-NAM-1_all_gene.bed\n",
      "130 NLRs found in gff file for Tzi8. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tzi8/genome/Zm-Tzi8-REFERENCE-NAM-1_NLR.bed\n",
      "M162W gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/M162W/genome/Zm-M162W-REFERENCE-NAM-1_all_gene.bed\n",
      "122 NLRs found in gff file for M162W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M162W/genome/Zm-M162W-REFERENCE-NAM-1_NLR.bed\n",
      "HP301 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/HP301/genome/Zm-HP301-REFERENCE-NAM-1_all_gene.bed\n",
      "116 NLRs found in gff file for HP301. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/HP301/genome/Zm-HP301-REFERENCE-NAM-1_NLR.bed\n",
      "CML277 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML277/genome/Zm-CML277-REFERENCE-NAM-1_all_gene.bed\n",
      "126 NLRs found in gff file for CML277. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML277/genome/Zm-CML277-REFERENCE-NAM-1_NLR.bed\n",
      "Ki3 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Ki3/genome/Zm-Ki3-REFERENCE-NAM-1_all_gene.bed\n",
      "129 NLRs found in gff file for Ki3. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki3/genome/Zm-Ki3-REFERENCE-NAM-1_NLR.bed\n"
     ]
    }
   ],
   "source": [
    "for i in gff3_files:\n",
    "    bed_generator(str(i), gene_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f8f06f24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B73 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_all_gene.bed\n",
      "126 NLRs found in gff file for B73. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_NLR.bed\n"
     ]
    }
   ],
   "source": [
    "bed_generator(gff3_files[10], gene_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1dea682",
   "metadata": {},
   "source": [
    "Take a look at the fulldata.txt files, which should have common gene names. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c09ca0cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "test='/global/scratch/users/chandlersutherland/e16/B73/genome/Zm00001eb.1.fulldata.txt'\n",
    "test_csv=pd.read_csv(test, sep='\\t', header=0, lineterminator='\\n', names=['gene_model', 'assembly_version', 'annotation_version', 'transcript_count', 'canonical_transcript_name', 'tandem_count', 'chr', 'gm_start', 'gm_end', 'locus_symbol', 'locus_name', 'gene_products', 'obo_terms', 'obo_names'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "a10b3900",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(test, 'r') as file: \n",
    "    file_contents=file.readlines()\n",
    "    for row in file_contents:\n",
    "        if row.find('Zm00001d023265') != -1:\n",
    "            print(row)\n",
    "#test_csv['locus_symbol']=test_csv['locus_symbol'].str.strip('         ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ba631804",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_model</th>\n",
       "      <th>assembly_version</th>\n",
       "      <th>annotation_version</th>\n",
       "      <th>transcript_count</th>\n",
       "      <th>canonical_transcript_name</th>\n",
       "      <th>tandem_count</th>\n",
       "      <th>chr</th>\n",
       "      <th>gm_start</th>\n",
       "      <th>gm_end</th>\n",
       "      <th>locus_symbol</th>\n",
       "      <th>locus_name</th>\n",
       "      <th>gene_products</th>\n",
       "      <th>obo_terms</th>\n",
       "      <th>obo_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>40647</td>\n",
       "      <td>Zm00001eb405930</td>\n",
       "      <td>Zm-B73-REFERENCE-NAM-5.0</td>\n",
       "      <td>Zm00001eb.1</td>\n",
       "      <td>2</td>\n",
       "      <td>Zm00001eb405930_T002</td>\n",
       "      <td>3</td>\n",
       "      <td>chr10</td>\n",
       "      <td>3333939</td>\n",
       "      <td>3335714</td>\n",
       "      <td>rp1</td>\n",
       "      <td>resistance to Puccinia sorghi1                ...</td>\n",
       "      <td>{\"nucleotide-binding site (NBS)-leucine-rich r...</td>\n",
       "      <td>{GO:0043531,GO:0005524,GO:0006952,GO:0043531,G...</td>\n",
       "      <td>{\"ADP binding\",\"ATP binding\",\"defense response...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_model          assembly_version  annotation_version  \\\n",
       "40647  Zm00001eb405930  Zm-B73-REFERENCE-NAM-5.0  Zm00001eb.1          \n",
       "\n",
       "       transcript_count  canonical_transcript_name  tandem_count       chr  \\\n",
       "40647                 2  Zm00001eb405930_T002                  3  chr10      \n",
       "\n",
       "       gm_start   gm_end locus_symbol  \\\n",
       "40647   3333939  3335714          rp1   \n",
       "\n",
       "                                              locus_name  \\\n",
       "40647  resistance to Puccinia sorghi1                ...   \n",
       "\n",
       "                                           gene_products  \\\n",
       "40647  {\"nucleotide-binding site (NBS)-leucine-rich r...   \n",
       "\n",
       "                                               obo_terms  \\\n",
       "40647  {GO:0043531,GO:0005524,GO:0006952,GO:0043531,G...   \n",
       "\n",
       "                                               obo_names  \n",
       "40647  {\"ADP binding\",\"ATP binding\",\"defense response...  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#test_csv.to_csv('/global/scratch/users/chandlersutherland/e16/B73/genome/gene_names.txt')\n",
    "test_csv[test_csv['gene_model']=='Zm00001eb405930']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8f0e64a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define a function that takes a gff3 file path as input, and outputs hv and non-hv NLR bed files \n",
    "def bed_generator_hv(gff_filepath, gene_table):\n",
    "    #read in and cleanup gff3 file \n",
    "    gff3=pd.read_csv(gff_filepath, sep='\\t', skiprows=3, header=0, lineterminator='\\n')\n",
    "    gff3=gff3.reset_index()\n",
    "    gff3=gff3.rename(columns={gff3.columns[0]:'chr', gff3.columns[1]:'source', gff3.columns[2]:'sequence_ontology', gff3.columns[3]:'start', gff3.columns[4]:'end', gff3.columns[5]:'score', gff3.columns[6]:'strand', gff3.columns[7]:'phase', gff3.columns[8]:'attributes'})\n",
    "\n",
    "    #get metadata\n",
    "    accession=gff_filepath.split('/')[6]\n",
    "    basename=gff_filepath.split('/')[8].split('.')[0]\n",
    "    print(accession+ ' gff3 file loaded')\n",
    "    \n",
    "    #filter gff3 to just genes, and expand attributes  \n",
    "    genes = gff3[gff3['sequence_ontology'] == 'gene']\n",
    "    attributes=genes.iloc[:,8].str.split(';',expand=True)\n",
    "    genes.loc[:,'ID']=attributes.iloc[:,0].str.split('=', expand=True)[1]\n",
    "    genes.loc[:,'biotype']=attributes.iloc[:,1].str.split('=', expand=True)[1]\n",
    "    genes.loc[:,'logic_name']=attributes.iloc[:,2].str.split('=', expand=True)[1]\n",
    "    \n",
    "    bed = genes.loc[:,['chr', 'start', 'end', 'ID', 'strand']]\n",
    "    bed = bed.rename(columns={'chr':'chrom', 'start':'chromStart', 'end':'chromEnd', 'ID':'name', 'strand':'strand'})\n",
    "    \n",
    "    #Load gene names\n",
    "    gene=pd.read_csv(gene_table,sep = '\\t')\n",
    "    nlr_df=gene[gene['Ecotype'] == accession.lower()]\n",
    "    nlrs=nlr_df.loc[:,'Gene'].str.split('_', expand=True).iloc[:,0].str.replace('ZM', 'Zm').str.replace('AB', 'ab').str.replace('EB','eb').unique()\n",
    "    \n",
    "    hv_nlrs=nlr_df[nlr_df['HV']==1].loc[:,'Gene'].str.split('_', expand=True).iloc[:,0].str.replace('ZM', 'Zm').str.replace('AB', 'ab').str.replace('EB','eb').unique()\n",
    "    nhv_nlrs=nlr_df[nlr_df['HV']==0].loc[:,'Gene'].str.split('_', expand=True).iloc[:,0].str.replace('ZM', 'Zm').str.replace('AB', 'ab').str.replace('EB','eb').unique()\n",
    "    #quality stats \n",
    "    match=bed['name'].isin(hv_nlrs).sum()\n",
    "    unmatch=len(hv_nlrs)-match \n",
    "    print(str(match)+' hvNLRs found in gff file for '+accession+'. We are missing '+str(unmatch)+'.')\n",
    "    match=bed['name'].isin(nhv_nlrs).sum()\n",
    "    unmatch=len(nhv_nlrs)-match \n",
    "    print(str(match)+' non-hvNLRs found in gff file for '+accession+'. We are missing '+str(unmatch)+'.')   \n",
    "    #write out NLR bed file\n",
    "    hv_nlr_bed=bed[bed['name'].isin(hv_nlrs)]\n",
    "    hv_nlr_bed.to_csv('/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_hvNLR.bed', sep='\\t', index=False)\n",
    "    print('NLR bed written to '+'/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_hvNLR.bed')\n",
    "    \n",
    "    nhv_nlr_bed=bed[bed['name'].isin(nhv_nlrs)]\n",
    "    nhv_nlr_bed.to_csv('/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_nhvNLR.bed', sep='\\t', index=False)\n",
    "    print('NLR bed written to '+'/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_nhvNLR.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ad02f173",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B97 gff3 file loaded\n",
      "12 hvNLRs found in gff file for B97. We are missing 0.\n",
      "113 non-hvNLRs found in gff file for B97. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B97/genome/Zm-B97-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B97/genome/Zm-B97-REFERENCE-NAM-1_nhvNLR.bed\n",
      "Oh43 gff3 file loaded\n",
      "11 hvNLRs found in gff file for Oh43. We are missing 0.\n",
      "120 non-hvNLRs found in gff file for Oh43. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh43/genome/Zm-Oh43-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh43/genome/Zm-Oh43-REFERENCE-NAM-1_nhvNLR.bed\n",
      "Tx303 gff3 file loaded\n",
      "11 hvNLRs found in gff file for Tx303. We are missing 0.\n",
      "110 non-hvNLRs found in gff file for Tx303. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tx303/genome/Zm-Tx303-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tx303/genome/Zm-Tx303-REFERENCE-NAM-1_nhvNLR.bed\n",
      "CML69 gff3 file loaded\n",
      "19 hvNLRs found in gff file for CML69. We are missing 0.\n",
      "114 non-hvNLRs found in gff file for CML69. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML69/genome/Zm-CML69-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML69/genome/Zm-CML69-REFERENCE-NAM-1_nhvNLR.bed\n",
      "CML333 gff3 file loaded\n",
      "22 hvNLRs found in gff file for CML333. We are missing 0.\n",
      "109 non-hvNLRs found in gff file for CML333. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML333/genome/Zm-CML333-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML333/genome/Zm-CML333-REFERENCE-NAM-1_nhvNLR.bed\n",
      "NC358 gff3 file loaded\n",
      "13 hvNLRs found in gff file for NC358. We are missing 0.\n",
      "119 non-hvNLRs found in gff file for NC358. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC358/genome/Zm-NC358-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC358/genome/Zm-NC358-REFERENCE-NAM-1_nhvNLR.bed\n",
      "NC350 gff3 file loaded\n",
      "17 hvNLRs found in gff file for NC350. We are missing 0.\n",
      "119 non-hvNLRs found in gff file for NC350. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC350/genome/Zm-NC350-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC350/genome/Zm-NC350-REFERENCE-NAM-1_nhvNLR.bed\n",
      "Mo18W gff3 file loaded\n",
      "10 hvNLRs found in gff file for Mo18W. We are missing 0.\n",
      "107 non-hvNLRs found in gff file for Mo18W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Mo18W/genome/Zm-Mo18W-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Mo18W/genome/Zm-Mo18W-REFERENCE-NAM-1_nhvNLR.bed\n",
      "CML103 gff3 file loaded\n",
      "13 hvNLRs found in gff file for CML103. We are missing 0.\n",
      "106 non-hvNLRs found in gff file for CML103. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML103/genome/Zm-CML103-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML103/genome/Zm-CML103-REFERENCE-NAM-1_nhvNLR.bed\n",
      "CML247 gff3 file loaded\n",
      "8 hvNLRs found in gff file for CML247. We are missing 0.\n",
      "106 non-hvNLRs found in gff file for CML247. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML247/genome/Zm-CML247-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML247/genome/Zm-CML247-REFERENCE-NAM-1_nhvNLR.bed\n",
      "B73 gff3 file loaded\n",
      "10 hvNLRs found in gff file for B73. We are missing 0.\n",
      "116 non-hvNLRs found in gff file for B73. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_nhvNLR.bed\n",
      "CML228 gff3 file loaded\n",
      "12 hvNLRs found in gff file for CML228. We are missing 0.\n",
      "114 non-hvNLRs found in gff file for CML228. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML228/genome/Zm-CML228-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML228/genome/Zm-CML228-REFERENCE-NAM-1_nhvNLR.bed\n",
      "P39 gff3 file loaded\n",
      "10 hvNLRs found in gff file for P39. We are missing 0.\n",
      "108 non-hvNLRs found in gff file for P39. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/P39/genome/Zm-P39-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/P39/genome/Zm-P39-REFERENCE-NAM-1_nhvNLR.bed\n",
      "Oh7B gff3 file loaded\n",
      "8 hvNLRs found in gff file for Oh7B. We are missing 0.\n",
      "113 non-hvNLRs found in gff file for Oh7B. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh7B/genome/Zm-Oh7B-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh7B/genome/Zm-Oh7B-REFERENCE-NAM-1_nhvNLR.bed\n",
      "MS71 gff3 file loaded\n",
      "14 hvNLRs found in gff file for MS71. We are missing 0.\n",
      "108 non-hvNLRs found in gff file for MS71. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/MS71/genome/Zm-Ms71-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/MS71/genome/Zm-Ms71-REFERENCE-NAM-1_nhvNLR.bed\n",
      "IL14H gff3 file loaded\n",
      "17 hvNLRs found in gff file for IL14H. We are missing 0.\n",
      "108 non-hvNLRs found in gff file for IL14H. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/IL14H/genome/Zm-Il14H-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/IL14H/genome/Zm-Il14H-REFERENCE-NAM-1_nhvNLR.bed\n",
      "Ki11 gff3 file loaded\n",
      "13 hvNLRs found in gff file for Ki11. We are missing 0.\n",
      "107 non-hvNLRs found in gff file for Ki11. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki11/genome/Zm-Ki11-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki11/genome/Zm-Ki11-REFERENCE-NAM-1_nhvNLR.bed\n",
      "CML52 gff3 file loaded\n",
      "6 hvNLRs found in gff file for CML52. We are missing 0.\n",
      "124 non-hvNLRs found in gff file for CML52. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML52/genome/Zm-CML52-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML52/genome/Zm-CML52-REFERENCE-NAM-1_nhvNLR.bed\n",
      "CML322 gff3 file loaded\n",
      "14 hvNLRs found in gff file for CML322. We are missing 0.\n",
      "107 non-hvNLRs found in gff file for CML322. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML322/genome/Zm-CML322-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML322/genome/Zm-CML322-REFERENCE-NAM-1_nhvNLR.bed\n",
      "KY21 gff3 file loaded\n",
      "17 hvNLRs found in gff file for KY21. We are missing 0.\n",
      "109 non-hvNLRs found in gff file for KY21. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/KY21/genome/Zm-Ky21-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/KY21/genome/Zm-Ky21-REFERENCE-NAM-1_nhvNLR.bed\n",
      "M37W gff3 file loaded\n",
      "18 hvNLRs found in gff file for M37W. We are missing 0.\n",
      "116 non-hvNLRs found in gff file for M37W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M37W/genome/Zm-M37W-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M37W/genome/Zm-M37W-REFERENCE-NAM-1_nhvNLR.bed\n",
      "Tzi8 gff3 file loaded\n",
      "15 hvNLRs found in gff file for Tzi8. We are missing 0.\n",
      "115 non-hvNLRs found in gff file for Tzi8. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tzi8/genome/Zm-Tzi8-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tzi8/genome/Zm-Tzi8-REFERENCE-NAM-1_nhvNLR.bed\n",
      "M162W gff3 file loaded\n",
      "13 hvNLRs found in gff file for M162W. We are missing 0.\n",
      "109 non-hvNLRs found in gff file for M162W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M162W/genome/Zm-M162W-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M162W/genome/Zm-M162W-REFERENCE-NAM-1_nhvNLR.bed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HP301 gff3 file loaded\n",
      "15 hvNLRs found in gff file for HP301. We are missing 0.\n",
      "101 non-hvNLRs found in gff file for HP301. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/HP301/genome/Zm-HP301-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/HP301/genome/Zm-HP301-REFERENCE-NAM-1_nhvNLR.bed\n",
      "CML277 gff3 file loaded\n",
      "11 hvNLRs found in gff file for CML277. We are missing 0.\n",
      "115 non-hvNLRs found in gff file for CML277. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML277/genome/Zm-CML277-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML277/genome/Zm-CML277-REFERENCE-NAM-1_nhvNLR.bed\n",
      "Ki3 gff3 file loaded\n",
      "17 hvNLRs found in gff file for Ki3. We are missing 0.\n",
      "112 non-hvNLRs found in gff file for Ki3. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki3/genome/Zm-Ki3-REFERENCE-NAM-1_hvNLR.bed\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki3/genome/Zm-Ki3-REFERENCE-NAM-1_nhvNLR.bed\n"
     ]
    }
   ],
   "source": [
    "for i in gff3_files:\n",
    "    bed_generator_hv(str(i), gene_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e7d7b4d",
   "metadata": {},
   "source": [
    "Repeat but with exons to create a bed file for methylation processing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5932dbba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define a function that takes a gff3 file path as input, and outputs an all exon bed file\n",
    "def bed_generator_exon(gff_filepath):\n",
    "    #read in and cleanup gff3 file \n",
    "    gff3=pd.read_csv(gff_filepath, sep='\\t', skiprows=3, header=0, lineterminator='\\n')\n",
    "    gff3=gff3.reset_index()\n",
    "    gff3=gff3.rename(columns={gff3.columns[0]:'chr', gff3.columns[1]:'source', gff3.columns[2]:'sequence_ontology', gff3.columns[3]:'start', gff3.columns[4]:'end', gff3.columns[5]:'score', gff3.columns[6]:'strand', gff3.columns[7]:'phase', gff3.columns[8]:'attributes'})\n",
    "\n",
    "    #get metadata\n",
    "    accession=gff_filepath.split('/')[6]\n",
    "    basename=gff_filepath.split('/')[8].split('.')[0]\n",
    "    print(accession+ ' gff3 file loaded')\n",
    "\n",
    "    exons = gff3[gff3['sequence_ontology'] == 'exon']\n",
    "\n",
    "    attributes=exons.iloc[:,8].str.split(';',expand=True)\n",
    "    exons.loc[:,'ID']=attributes.iloc[:,0].str.split('=', expand=True)[1]\n",
    "    exons.loc[:,'biotype']=attributes.iloc[:,1].str.split('=', expand=True)[1]\n",
    "    exons.loc[:,'logic_name']=attributes.iloc[:,2].str.split('=', expand=True)[1]\n",
    "\n",
    "    bed = exons.loc[:,['chr', 'start', 'end', 'ID', 'biotype', 'strand']]\n",
    "    bed = bed.rename(columns={'chr':'chrom', 'start':'chromStart', 'end':'chromEnd', 'ID':'name', 'biotype':'exon', 'strand':'strand'})\n",
    "\n",
    "    bed.to_csv('/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_all_exon.bed', sep='\\t')\n",
    "    print('all exon bed written to '+'/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_all_exon.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "942cce8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B97 gff3 file loaded\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/B97/genome/Zm-B97-REFERENCE-NAM-1_all_exon.bed\n",
      "Oh43 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/Oh43/genome/Zm-Oh43-REFERENCE-NAM-1_all_exon.bed\n",
      "Tx303 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/Tx303/genome/Zm-Tx303-REFERENCE-NAM-1_all_exon.bed\n",
      "CML69 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML69/genome/Zm-CML69-REFERENCE-NAM-1_all_exon.bed\n",
      "CML333 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML333/genome/Zm-CML333-REFERENCE-NAM-1_all_exon.bed\n",
      "KI11 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/KI11/genome/Zm-Ki11-REFERENCE-NAM-1_all_exon.bed\n",
      "NC358 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/NC358/genome/Zm-NC358-REFERENCE-NAM-1_all_exon.bed\n",
      "TZI8 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/TZI8/genome/Zm-Tzi8-REFERENCE-NAM-1_all_exon.bed\n",
      "NC350 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/NC350/genome/Zm-NC350-REFERENCE-NAM-1_all_exon.bed\n",
      "TX303 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/TX303/genome/Zm-Tx303-REFERENCE-NAM-1_all_exon.bed\n",
      "OH43 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/OH43/genome/Zm-Oh43-REFERENCE-NAM-1_all_exon.bed\n",
      "KI3 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/KI3/genome/Zm-Ki3-REFERENCE-NAM-1_all_exon.bed\n",
      "Mo18W gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/Mo18W/genome/Zm-Mo18W-REFERENCE-NAM-1_all_exon.bed\n",
      "CML103 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML103/genome/Zm-CML103-REFERENCE-NAM-1_all_exon.bed\n",
      "CML247 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML247/genome/Zm-CML247-REFERENCE-NAM-1_all_exon.bed\n",
      "B73 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_all_exon.bed\n",
      "CML228 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML228/genome/Zm-CML228-REFERENCE-NAM-1_all_exon.bed\n",
      "P39 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/P39/genome/Zm-P39-REFERENCE-NAM-1_all_exon.bed\n",
      "Oh7B gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/Oh7B/genome/Zm-Oh7B-REFERENCE-NAM-1_all_exon.bed\n",
      "MS71 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/MS71/genome/Zm-Ms71-REFERENCE-NAM-1_all_exon.bed\n",
      "MO18W gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/MO18W/genome/Zm-Mo18W-REFERENCE-NAM-1_all_exon.bed\n",
      "IL14H gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/IL14H/genome/Zm-Il14H-REFERENCE-NAM-1_all_exon.bed\n",
      "Ki11 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/Ki11/genome/Zm-Ki11-REFERENCE-NAM-1_all_exon.bed\n",
      "CML52 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML52/genome/Zm-CML52-REFERENCE-NAM-1_all_exon.bed\n",
      "OH7B gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/OH7B/genome/Zm-Oh7B-REFERENCE-NAM-1_all_exon.bed\n",
      "CML322 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML322/genome/Zm-CML322-REFERENCE-NAM-1_all_exon.bed\n",
      "KY21 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/KY21/genome/Zm-Ky21-REFERENCE-NAM-1_all_exon.bed\n",
      "M37W gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/M37W/genome/Zm-M37W-REFERENCE-NAM-1_all_exon.bed\n",
      "Tzi8 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/Tzi8/genome/Zm-Tzi8-REFERENCE-NAM-1_all_exon.bed\n",
      "M162W gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/M162W/genome/Zm-M162W-REFERENCE-NAM-1_all_exon.bed\n",
      "HP301 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/HP301/genome/Zm-HP301-REFERENCE-NAM-1_all_exon.bed\n",
      "CML277 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/CML277/genome/Zm-CML277-REFERENCE-NAM-1_all_exon.bed\n",
      "Ki3 gff3 file loaded\n",
      "all exon bed written to /global/scratch/users/chandlersutherland/e16/Ki3/genome/Zm-Ki3-REFERENCE-NAM-1_all_exon.bed\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for i in gff3_files:\n",
    "    bed_generator_exon(str(i))\n",
    "    \n",
    "exon_files=glob.glob('/global/scratch/users/chandlersutherland/e16/*/genome/*_all_exon.bed')\n",
    "len(exon_files)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
